File: 0046 Permutations/leet_On_On.go
func permute(nums []int) [][]int {
    out := [][]int{}
    used := make([]bool, len(nums))

    var backtrack func(curr []int)

    backtrack = func (curr []int) {
        if len(curr) == len(nums) {
            cp := make([]int, len(curr))
            copy(cp, curr)
            out = append(out, cp)
            return
        }
        // O(n)
        for n := 0; n < len(nums); n++ {
            if used[n] {
                continue
            }
            used[n] = true
            curr = append(curr, nums[n])
            backtrack(curr)
            curr = curr[:len(curr)-1]
            used[n] = false
        }
    }

    backtrack([]int{})

    return out
}

File: 0046 Permutations/leet_On_On.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        out = []
        # , O(n)
        used = [False] * len(nums)

        def backtrack(curr):
            if len(curr) == len(nums):
                out.append(curr[:])
                return
            # O(n)
            for n in range(len(nums)):
                # O(n)
                if not used[n]:
                    used[n] = True
                    curr.append(nums[n])
                    backtrack(curr)
                    curr.pop()
                    used[n] = False

        backtrack([])

        return out
